apps/astarte_realm_management/lib/astarte_realm_management/device_removal/device_remover.ex

```diff
@@ -0,0 +1,166 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2023 SECO Mind Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.RealmManagement.DeviceRemoval.DeviceRemover do
+  @moduledoc """
+  This module handles data deletion for a device using a Task.
+  The Task may fail at any time, notably if the database is not
+  available.
+  See Astarte.RealmManagement.DeviceRemoval.Scheduler for handling failures.
+  """
+
+  use Task
+  require Logger
+  alias Astarte.Core.InterfaceDescriptor
+  alias Astarte.Core.CQLUtils
+  alias Astarte.RealmManagement.Queries
+  alias Astarte.Core.Device
+
+  @spec run(%{:device_id => <<_::128>>, :realm_name => binary()}) :: :ok | no_return()
+  def run(%{realm_name: realm_name, device_id: device_id}) do
+    encoded_device_id = Device.encode_device_id(device_id)
+    _ = Logger.info("Starting to remove device #{encoded_device_id}", tag: "device_delete_start")
+
+    datastream_keys = Queries.retrieve_individual_datastreams_keys!(realm_name, device_id)
+    :ok = delete_individual_datastreams!(realm_name, datastream_keys)
+    property_keys = Queries.retrieve_individual_properties_keys!(realm_name, device_id)
+    :ok = delete_individual_properties!(realm_name, property_keys)
+    introspection = retrieve_device_introspection_map!(realm_name, device_id)
+    object_interfaces = filter_object_interfaces!(realm_name, introspection)
+    object_tables = Enum.map(object_interfaces, &object_interface_to_table_name/1)
+
```

Comment:
Sometimes directly calling `Enum` functions and some other times defining private helpers to call them, I'd prefer a consistent behavior. I'd suggest moving the `Enum.each` here and leaving as helper only the function that unpacks the struct and uses that to create the arguments for the function call.
If you combine this with dropping the match on `:ok`, you can easily group functions in pipes which end up handling a specific side effect, e.g:
```elixir
# Delete individual datastreams
Queries.retrieve_individual_datastreams_keys!(realm_name, device_id)
|> Enum.each(&delete_individual_datastreams_from_key!/1)

...
end)

defp delete_individual_datastreams_from_key!(key) do
  %{
    device_id: device_id,
    interface_id: interface_id,
    endpoint_id: endpoint_id,
    path: path
  } = key

  Queries.delete_individual_datastream_values!(
    realm_name,
    device_id,
    interface_id,
    endpoint_id,
    path
  )
end
```